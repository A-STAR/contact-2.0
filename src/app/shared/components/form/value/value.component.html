<ng-container [ngSwitch]="value && value.typeCode">

  <!-- Number -->
  <input
    *ngSwitchCase="1"
    type="number"
    class="form-control"
    [ngModel]="value.valueN"
    (ngModelChange)="onNumberChange($event)"
  />

  <!-- Date -->
  <app-datepicker
    *ngSwitchCase="2"
    [ngModel]="date"
    (ngModelChange)="onDateChange($event)">
  </app-datepicker>

  <!-- String -->
  <input
    *ngSwitchCase="3"
    type="text"
    class="form-control"
    [ngModel]="value.valueS"
    (ngModelChange)="onStringChange($event)"
  />

  <!-- Boolean -->
  <app-select
    *ngSwitchCase="4"
    [dictCode]="2"
    [ngModel]="value.valueB"
    (ngModelChange)="onBooleanChange($event)">
  </app-select>

  <!-- Float -->
  <input
    *ngSwitchCase="5"
    type="number"
    class="form-control"
    [ngModel]="value.valueN"
    (ngModelChange)="onNumberChange($event)"
  />

  <!-- Dictionary -->
  <app-select
    *ngSwitchCase="6"
    [dictCode]="value.dictNameCode"
    [ngModel]="value.valueN"
    (ngModelChange)="onNumberChange($event)">
  </app-select>

  <!-- DateTime -->
  <app-datetimepicker
    *ngSwitchCase="7"
    [ngModel]="date"
    (ngModelChange)="onDateChange($event)">
  </app-datetimepicker>
</ng-container>
