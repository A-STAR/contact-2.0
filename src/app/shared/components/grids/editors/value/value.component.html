<ng-container [ngSwitch]="type">

  <div *ngSwitchCase="1" class="text-right">
    <app-number
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    >
    </app-number>
  </div>

  <div *ngSwitchCase="2">
    <div style="padding: 10px;">
      <app-date
        [ngModel]="value"
        (ngModelChange)="onValueChange($event)"
      ></app-date>
    </div>
  </div>

  <div *ngSwitchCase="3">
    <app-text
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    >
    </app-text>
  </div>

  <div *ngSwitchCase="4">
    <app-select
      [dictCode]="2"
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    ></app-select>
  </div>

  <div *ngSwitchCase="5" class="text-right">
    <app-number
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    >
    </app-number>
  </div>

  <div *ngSwitchCase="6">
    <app-select
      [dictCode]="dictCode"
      [lookupKey]="lookupKey"
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    ></app-select>
  </div>

  <div *ngSwitchCase="7">
    <app-grid-datetime-edit
      [ngModel]="value"
      (ngModelChange)="onValueChange($event)"
    >
    </app-grid-datetime-edit>
  </div>

</ng-container>
