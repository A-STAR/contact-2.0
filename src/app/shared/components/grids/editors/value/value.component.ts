import { Component } from '@angular/core';
import { ICellEditorParams } from 'ag-grid/main';
import { ICellEditorAngularComp } from 'ag-grid-angular';

import { TYPE_CODES } from '@app/core/utils';

@Component({
  selector: 'app-grid-value-edit',
  templateUrl: './value.component.html',
})
export class ValueEditorComponent implements ICellEditorAngularComp {
  private params: ICellEditorParams;

  value: any;
  dictCode: number;

  refresh(): boolean {
    return false;
  }

  get type(): number {
    const { node, valueTypeKey } = this.params as any;
    return node.data[valueTypeKey] || TYPE_CODES.STRING;
  }

  agInit(params: ICellEditorParams): void {
    const { node, dictCode, value } = params as any;
    this.params = params;
    this.value = value;
    this.dictCode = typeof dictCode === 'function' ? dictCode(node.data) : dictCode;
  }

  getValue(): any {
    return this.value;
  }

  isPopup(): boolean {
    return true;
  }

  onValueChange(value: any): void {
    this.value = value;
  }
}
