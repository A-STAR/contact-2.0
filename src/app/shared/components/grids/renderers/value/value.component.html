<ng-container [ngSwitch]="type">

  <div *ngSwitchCase="1" class="text-right">
    {{ value }}
  </div>

  <div *ngSwitchCase="2">
    {{ value | momentFormat:'L' }}
  </div>

  <div *ngSwitchCase="3">
    {{ value }}
  </div>

  <div *ngSwitchCase="4">
    <!-- TODO(d.maltsev): use dictionary 2 instead -->
    {{ (value ? 'default.boolean.TRUE' : 'default.boolean.FALSE') | translate }}
  </div>

  <div *ngSwitchCase="5" class="text-right">
    {{ value }}
  </div>

  <div *ngSwitchCase="6">
    <!-- TODO(d.maltsev): add dict rendering if it's used anywhere -->
    {{ value }}
  </div>

  <div *ngSwitchCase="7">
    {{ value | momentFormat:'L LTS' }}
  </div>
</ng-container>
