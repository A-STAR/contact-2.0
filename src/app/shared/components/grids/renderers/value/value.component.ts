import { ChangeDetectionStrategy, Component, ChangeDetectorRef } from '@angular/core';
import { ICellRendererAngularComp } from 'ag-grid-angular';
import { map } from 'rxjs/operators';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import { IValueRendererParams } from '@app/shared/components/grids/grids.interface';

import { UserDictionariesService } from '@app/core/user/dictionaries/user-dictionaries.service';

import { TYPE_CODES, getRawValue } from '@app/core/utils';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'app-grid-tick-renderer',
  templateUrl: './value.component.html',
})
export class ValueRendererComponent implements ICellRendererAngularComp {

  type: TYPE_CODES;
  value: any;
  dictValue$: Observable<string>;

  constructor(
    private cdRef: ChangeDetectorRef,
    private userDictionariesService: UserDictionariesService,
  ) {}

  agInit(params: IValueRendererParams): void {
    const { data, valueTypeKey, value } = params;
    this.type = data[valueTypeKey];
    // value is undefined on initial load
    // value is present after cellEditor changes
    if (this.type) {
      this.value = value ? value : getRawValue(params.data, valueTypeKey);
    } else if (!this.type) {
      this.value = value;
      this.type = TYPE_CODES.STRING;
    }

    if (this.type === TYPE_CODES.DICT) {
      this.dictValue$ = this.getDictValue(params, this.value);
    }
    // we mutate data because cellEditor won't have initial value
    // TODO(i.lobanov): find a way to avoid it
    data.value = this.value;

    this.cdRef.markForCheck();
  }

  refresh(): boolean {
    return false;
  }
  private getDictValue(params: IValueRendererParams, value: any): Observable<string> {
    const { data, valueTypeParams } = params;
    // TODO(i.lobanov): lookupKey
    if (valueTypeParams && valueTypeParams.dictCode) {

      const code = typeof valueTypeParams.dictCode === 'function' ? valueTypeParams.dictCode(data)
        : valueTypeParams.dictCode;

      return this.userDictionariesService
        .getDictionary(code)
        .pipe(
          map(terms => {
            const term = terms.find(t => t.code === value);
            return term ? term.name : value;
          }),
        );
    }
    return of(value);
  }
}
