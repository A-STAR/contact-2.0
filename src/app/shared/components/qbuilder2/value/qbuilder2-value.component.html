<span class="value">
  <span *ngIf="displayAsArray">(</span>

  <ng-container *ngFor="let v of value; let i = index; let last = last">
    <ng-container [ngSwitch]="controlType">

      <!-- Datepicker -->
      <ng-container *ngSwitchCase="'date'">
        <app-input-datepicker
          buttonClass="btn btn-sm btn-default"
          inputClass="form-control input-sm"
          [ngModel]="toDate(v)">
        </app-input-datepicker>
      </ng-container>

      <!-- Dropdown -->
      <ng-container *ngSwitchCase="'set'">
        <app-select
          [ngModel]="v"
          [options]="options">
        </app-select>
      </ng-container>

      <!-- Default input -->
      <ng-container *ngSwitchDefault>
        <input
          type="text"
          class="form-control input-sm"
          [ngModel]="v"
          (change)="onValueChange($event, i)"
        />
      </ng-container>
    </ng-container>
    <span *ngIf="!last && displayAsArray">,&nbsp;</span>
  </ng-container>

  <span *ngIf="displayAsArray">)</span>

  <ng-container *ngIf="displayButtons">
    <button class="btn btn-xs btn-default" (click)="addValue()">+</button>
    <button class="btn btn-xs btn-default" (click)="removeValue()">-</button>
  </ng-container>
<span>
