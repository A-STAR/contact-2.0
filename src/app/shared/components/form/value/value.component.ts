import { ChangeDetectionStrategy, ChangeDetectorRef, Component, forwardRef } from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

import { ValueConverterService } from '../../../../core/converter/value-converter.service';

import { IValue } from '../../../../core/utils/value';

@Component({
  selector: 'app-value-input',
  templateUrl: './value.component.html',
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => ValueInputComponent),
      multi: true
    }
  ],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ValueInputComponent implements ControlValueAccessor {
  private _value: IValue;

  constructor(
    private cdRef: ChangeDetectorRef,
    private valueConverterService: ValueConverterService,
  ) {}

  // For some reason getter doesn't work
  date = null;

  writeValue(value: IValue): void {
    this._value = value;
    if (value && value.valueD) {
      this.date = this.valueConverterService.fromISO(value.valueD);
    }
    this.cdRef.markForCheck();
  }

  registerOnChange(fn: Function): void {
    this.propagateChange = fn;
  }

  registerOnTouched(): void {
  }

  get value(): IValue {
    return this._value;
  }

  onNumberChange(value: number): void {
    this._value.valueN = Array.isArray(value) ? value[0].value : value;
    this.propagateChange(this._value);
  }

  onStringChange(value: string): void {
    this._value.valueS = value;
    this.propagateChange(this._value);
  }

  onBooleanChange(value: number): void {
    this._value.valueB = value;
    this.propagateChange(this._value);
  }

  onDateChange(date: Date): void {
    this.date = date;
    this._value.valueD = this.valueConverterService.toISO(date);
    this.propagateChange(this._value);
  }

  private propagateChange: Function = () => {};
}
