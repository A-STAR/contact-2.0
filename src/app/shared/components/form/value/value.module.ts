import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { DatePickerModule } from '../datepicker/datepicker.module';
import { DateTimeModule } from '../datetime/datetime.module';
import { SelectModule } from '../select/select.module';

import { ValueInputComponent } from './value.component';

@NgModule({
  imports: [
    CommonModule,
    DatePickerModule,
    DateTimeModule,
    FormsModule,
    ReactiveFormsModule,
    SelectModule,
  ],
  exports: [
    ValueInputComponent,
  ],
  declarations: [
    ValueInputComponent,
  ],
})
export class ValueInputModule { }
