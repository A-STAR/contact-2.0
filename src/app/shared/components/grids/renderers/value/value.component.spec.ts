import { async as Async, ComponentFixture, TestBed } from '@angular/core/testing';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import { MomentModule } from '@app/shared/pipes/moment/moment.module';

import { ValueRendererComponent } from './value.component';

class TranslateLoaderMock {
  getTranslation(language: string): Observable<any> {
    return of([]);
  }
}

describe('ValueRendererComponent', () => {
  let fixture: ComponentFixture<ValueRendererComponent>;

  beforeEach(Async(() => {
    TestBed
      .configureTestingModule({
        imports: [
          MomentModule,
          TranslateModule.forRoot({
            loader: {
              provide: TranslateLoader,
              useClass: TranslateLoaderMock,
            },
          }),
        ],
        declarations: [
          ValueRendererComponent,
        ],
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ValueRendererComponent);
  });

  it('should render value renderer with typeCode 1 (number)', () => {
    fixture.componentInstance.agInit({
      value: 0,
      data: { typeCode: 1 },
      valueTypeKey: 'typeCode',
    } as any);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render value renderer with typeCode 3 (string)', () => {
    fixture.componentInstance.agInit({
      value: 'value',
      data: { typeCode: 3 },
      valueTypeKey: 'typeCode',
    } as any);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render value renderer with typeCode 4 (boolean) and value = false', () => {
    fixture.componentInstance.agInit({
      value: false,
      data: { typeCode: 4 },
      valueTypeKey: 'typeCode',
    } as any);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render value renderer with typeCode 4 (boolean) and value = true', () => {
    fixture.componentInstance.agInit({
      value: true,
      data: { typeCode: 4 },
      valueTypeKey: 'typeCode',
    } as any);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
