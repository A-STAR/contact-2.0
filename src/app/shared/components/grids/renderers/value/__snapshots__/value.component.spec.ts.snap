// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueRendererComponent should render value renderer with typeCode 1 (number) 1`] = `
<div
  id="root0"
  ng-version="5.2.4"
>
  
  

  
  
  <div
    class="text-right"
  >
    
    
  
  </div>
  

  
  
  

  
  
  

  
  
  

  
  
  

  
  
  

  
  
  

  

</div>
`;

exports[`ValueRendererComponent should render value renderer with typeCode 3 (string) 1`] = `
<div
  id="root1"
  ng-version="5.2.4"
>
  
  

  
  
  

  
  
  

  
  
  <div>
    
    value
  
  </div>
  

  
  
  

  
  
  

  
  
  

  
  
  

  

</div>
`;

exports[`ValueRendererComponent should render value renderer with typeCode 4 (boolean) and value = 0 1`] = `
<div
  id="root2"
  ng-version="5.2.4"
>
  
  

  
  
  

  
  
  

  
  
  

  
  
  <div>
    
    
    
    default.boolean.FALSE
  
  </div>
  

  
  
  

  
  
  

  
  
  

  

</div>
`;

exports[`ValueRendererComponent should render value renderer with typeCode 4 (boolean) and value = 1 1`] = `
<div
  id="root3"
  ng-version="5.2.4"
>
  
  

  
  
  

  
  
  

  
  
  

  
  
  <div>
    
    
    
    default.boolean.TRUE
  
  </div>
  

  
  
  

  
  
  

  
  
  

  

</div>
`;
